From 3e6a365d2ac9988058faef27629817908961422a Mon Sep 17 00:00:00 2001
From: Val Packett <val@invisiblethingslab.com>
Date: Fri, 28 Aug 2026 18:24:00 -0300
Subject: [PATCH] drm/virtio: pass the current ctx_id during PRIME import

The VIRTIO_GPU_CMD_RESOURCE_CREATE_BLOB command needs to have the
current context ID set in order for the VMM to handle it correctly.

Unfortunately the gem_prime_import callback does not have access to the
drm_file of the current DRM node which contains the ID, so define a
prime_fd_to_handle to store it in the driver's private state (protected
by a lock) before continuing on to the code that calls into the
gem_prime_import which can now pick it up.

Signed-off-by: Val Packett <val@invisiblethingslab.com>
---
 drivers/gpu/drm/virtio/virtgpu_drv.c   |  1 +
 drivers/gpu/drm/virtio/virtgpu_drv.h   |  6 ++++++
 drivers/gpu/drm/virtio/virtgpu_prime.c | 12 ++++++++++++
 3 files changed, 19 insertions(+)

diff --git a/drivers/gpu/drm/virtio/virtgpu_drv.c b/drivers/gpu/drm/virtio/virtgpu_drv.c
index 9650cf560b8e..db8a9efdf59d 100644
--- a/drivers/gpu/drm/virtio/virtgpu_drv.c
+++ b/drivers/gpu/drm/virtio/virtgpu_drv.c
@@ -344,6 +344,7 @@ static const struct drm_driver driver = {
 #if defined(CONFIG_DEBUG_FS)
 	.debugfs_init = virtio_gpu_debugfs_init,
 #endif
+	.prime_fd_to_handle = virtgpu_prime_fd_to_handle,
 	.gem_prime_import = virtgpu_gem_prime_import,
 	.gem_prime_import_sg_table = virtgpu_gem_prime_import_sg_table,
 
diff --git a/drivers/gpu/drm/virtio/virtgpu_drv.h b/drivers/gpu/drm/virtio/virtgpu_drv.h
index 2602992d9b2e..e0a004d5f851 100644
--- a/drivers/gpu/drm/virtio/virtgpu_drv.h
+++ b/drivers/gpu/drm/virtio/virtgpu_drv.h
@@ -306,6 +306,9 @@ struct virtio_gpu_device {
 
 	struct notifier_block pm_nb;
 
+	struct mutex prime_import_lock;
+	struct drm_file *prime_import_file_priv;
+
 	/* protects uuid state when exporting */
 	spinlock_t resource_export_lock;
 	/* protects map state and host_visible_mm */
@@ -532,6 +535,9 @@ int virtio_gpu_resource_assign_uuid(struct virtio_gpu_device *vgdev,
 				    struct virtio_gpu_object *bo);
 struct dma_buf *virtgpu_gem_prime_export(struct drm_gem_object *obj,
 					 int flags);
+int virtgpu_prime_fd_to_handle(struct drm_device *dev,
+			   struct drm_file *file_priv,
+			   int fd, u32 *handle);
 struct drm_gem_object *virtgpu_gem_prime_import(struct drm_device *dev,
 						struct dma_buf *buf);
 struct drm_gem_object *virtgpu_gem_prime_import_sg_table(
diff --git a/drivers/gpu/drm/virtio/virtgpu_prime.c b/drivers/gpu/drm/virtio/virtgpu_prime.c
index e5d662ac9644..a6730bb9d061 100644
--- a/drivers/gpu/drm/virtio/virtgpu_prime.c
+++ b/drivers/gpu/drm/virtio/virtgpu_prime.c
@@ -232,6 +232,7 @@ static int virtgpu_dma_buf_init_obj(struct drm_device *dev,
 	struct virtio_gpu_device *vgdev = dev->dev_private;
 	struct virtio_gpu_object_params params = { 0 };
 	struct dma_resv *resv = attach->dmabuf->resv;
+	struct virtio_gpu_fpriv *vfpriv = vgdev->prime_import_file_priv->driver_priv;
 	struct virtio_gpu_mem_entry *ents = NULL;
 	unsigned int nents;
 	int ret;
@@ -252,6 +253,7 @@ static int virtgpu_dma_buf_init_obj(struct drm_device *dev,
 	if (ret)
 		goto err_import;
 
+	params.ctx_id = vfpriv->ctx_id;
 	params.blob = true;
 	params.blob_mem = VIRTGPU_BLOB_MEM_GUEST;
 	params.blob_flags = VIRTGPU_BLOB_FLAG_USE_SHAREABLE;
@@ -336,6 +338,16 @@ static const struct dma_buf_attach_ops virtgpu_dma_buf_attach_ops = {
 	.invalidate_mappings = virtgpu_dma_buf_move_notify
 };
 
+int virtgpu_prime_fd_to_handle(struct drm_device *dev,
+			   struct drm_file *file_priv,
+			   int fd, u32 *handle)
+{
+	struct virtio_gpu_device *vgdev = dev->dev_private;
+	guard(mutex)(&vgdev->prime_import_lock);
+	vgdev->prime_import_file_priv = file_priv;
+	return drm_gem_prime_fd_to_handle(dev, file_priv, fd, handle);
+}
+
 struct drm_gem_object *virtgpu_gem_prime_import(struct drm_device *dev,
 						struct dma_buf *buf)
 {
